## Importing packages + reading in the data

In [ ]:
# 1. Load the Foreign Gifts data.
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_csv('ForeignGifts_edu.csv', low_memory=False)
df.head()

,ID,OPEID,Institution Name,City,State,Foreign Gift Received Date,Foreign Gift Amount,Gift Type,Country of Giftor,Giftor Name
0,1,102000,Jacksonville State University,Jacksonville,AL,43738,250000,Monetary Gift,CHINA,NaN
1,2,104700,Troy University,Troy,AL,43592,463657,Contract,CHINA,Confucius Institute Headquarters
2,3,105100,University of Alabama,Tuscaloosa,AL,43466,3649107,Contract,ENGLAND,Springer Nature Customer Service Ce
3,4,105100,University of Alabama,Tuscaloosa,AL,43472,1000,Contract,SAUDI ARABIA,Saudi Arabia Education Mission
4,5,105100,University of Alabama,Tuscaloosa,AL,43479,49476,Contract,SAUDI ARABIA,Saudi Arabia Education Mission


In [11]:
import sys
print(sys.version)
print(sys.executable)

3.14.7 | packaged by conda-forge | (main, Aug 19 2026, 15:29:14) [Clang 20.1.8 ]
/opt/miniconda3/envs/ds6001/bin/python


2. Describe the differences between different classes of gift types.

Contracts are the most common type and contribute the most money overall. Monetary gifts are less frequent and have a lower mean and median. Real-estate gifts are extremely rare (11 records), but their mean and median are the largest, so their typical recorded value is higher.

3. Answer the following questions:

<div class="alert alert-block alert-warning">
<b>Question 1:</b> Which country gives the most money, in total?
</div>

In [2]:
# Add the gift amounts for each country.
country_totals = df.groupby('Country of Giftor')['Foreign Gift Amount'].sum()
country_totals = country_totals.sort_values(ascending=False)
country_totals.head(10)

Country of Giftor
QATAR           2706240869
ENGLAND         1464906771
CHINA           1237952112
SAUDI ARABIA    1065205930
BERMUDA          899593972
CANADA           898160656
HONG KONG        887402529
JAPAN            655954776
SWITZERLAND      619899445
INDIA            539556490
Name: Foreign Gift Amount, dtype: int64

**Answer:** Qatar gives the most money in total, with **$2,706,240,869**.

<div class="alert alert-block alert-warning">
<b>Question 2:</b> Which country initiates the most gifts, by count?
</div>

In [3]:
# Count how many rows belong to each country.
country_counts = df['Country of Giftor'].value_counts()
country_counts.head(10)

Country of Giftor
ENGLAND         3655
CHINA           2461
CANADA          2344
JAPAN           1896
SWITZERLAND     1676
SAUDI ARABIA    1610
FRANCE          1437
GERMANY         1394
HONG KONG       1080
SOUTH KOREA      811
Name: count, dtype: int64

**Answer:** England initiates the most gifts, with **3,655** records.

<div class="alert alert-block alert-warning">
<b>Question 3:</b> Which country gives the largest gifts, on average?
</div>

In [4]:
# Find the average gift for each country.
country_average = df.groupby('Country of Giftor')['Foreign Gift Amount'].mean()
country_average = country_average.sort_values(ascending=False)
country_average.head(10)

Country of Giftor
BERMUDA     7.688837e+06
ECUADOR     3.985140e+06
QATAR       3.905109e+06
ANTIGUA     3.435545e+06
JERSEY      3.064589e+06
MOROCCO     2.740214e+06
RUSSIA      2.025462e+06
GUERNSEY    1.750000e+06
IRAQ        1.665520e+06
CYPRUS      1.562512e+06
Name: Foreign Gift Amount, dtype: float64

**Answer:** Bermuda has the largest average gift, approximately **$7.69 million**. Counts are shown because averages for countries with few gifts can be unstable.

## Ordering by institution

<div class="alert alert-block alert-warning">
<b>Question 4:</b> Which institution receive the most money, in total?
</div>

<div class="alert alert-block alert-warning">
<b>Question 5:</b> Which institution receive the greatest number of gifts, by count?
</div>

In [5]:
# Add the gift amounts for each institution.
institution_totals = df.groupby('Institution Name')['Foreign Gift Amount'].sum()
institution_totals = institution_totals.sort_values(ascending=False)

# Count the number of gifts for each institution.
institution_counts = df['Institution Name'].value_counts()

print(institution_totals.head(1))
print(institution_counts.head(1))

Institution Name
Carnegie Mellon University    1477922504
Name: Foreign Gift Amount, dtype: int64
Institution Name
University of California, Los Angeles    3916
Name: count, dtype: int64


**Answers:** Carnegie Mellon University receives the most money (**$1,477,922,504**). The University of California, Los Angeles receives the most gifts (**3,916**).

## Comparing summary statistics

<div class="alert alert-block alert-warning">
<b>Question 6:</b> What is the average amount each university receives? Median amount? If they are different, why?
</div>

In [6]:
# Use the total received by each university.
amount_per_university = institution_totals
mean_received = amount_per_university.mean()
median_received = amount_per_university.median()
print(f'Mean received per university:   ${mean_received:,.2f}')
print(f'Median received per university: ${median_received:,.2f}')


Mean received per university:   $52,202,878.97
Median received per university: $6,486,791.50


**Answer:** The mean is **$52,202,878.97** per university and the median is **$6,486,791.50**. The distribution is strongly skewed.  Some universities receive extremely large ammounts, bringing the mean much higher, while the median is less affected by  outliers.

## Relationship between variables
<div class="alert alert-block alert-warning">
<b>Question 7:</b> What is the largest flow of a single country to a single institution?
</div>



In [7]:
# Make a table of money sent from each country to each institution.
country_institution = pd.crosstab(
    df['Country of Giftor'],
    df['Institution Name'],
    values=df['Foreign Gift Amount'],
    aggfunc='sum'
)
country_institution.head()

Institution Name,Adelphi University,Albert Einstein College of Medicine,Alfred University,American University (The),Amherst College,Arizona State University,Auburn University Montgomery,Babson College,Ball State University,Barnard College,...,Wilkes University,William Marshall Rice University,Williams College,Winthrop University,Worcester Polytechnic Institute,Wright State University,Xavier University of Louisiana,Yale University,Yeshiva University,Young Americans College of the Performing Arts (The)
Country of Giftor,,,,,,,,,,,,,,,,,,,,,
AFGHANISTAN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
AMERICAN SAMOA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ANGOLA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ANTIGUA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ARGENTINA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [8]:
# Turn the table into one long list and sort it.
flows = country_institution.stack().sort_values(ascending=False)
flows.head(10)

Country of Giftor     Institution Name              
QATAR                 Cornell University                1.018473e+09
BERMUDA               Carnegie Mellon University        7.500000e+08
QATAR                 Texas A&M University              5.040994e+08
                      Carnegie Mellon University        4.254107e+08
                      Georgetown University             3.271283e+08
                      Northwestern University           3.111486e+08
CANADA                Brigham Young University          2.599278e+08
UNITED ARAB EMIRATES  University of Colorado Boulder    2.197348e+08
ENGLAND               Harvard University                1.883283e+08
HONG KONG             Yale University                   1.611312e+08
dtype: float64

**Answer:** The largest country-to-institution flow is **Qatar to Cornell University**, totaling **$1,018,473,315**.

### Flow of information using alluvial plots

<div class="alert alert-block alert-warning">
<b>Question 8:</b> Which are the top giftors to US academic institutions?
</div>

In [9]:
# Remove rows where the giftor name is missing.
known_giftors = df.dropna(subset=['Giftor Name'])

# Find the giftors who gave the most money.
giftor_totals = known_giftors.groupby('Giftor Name')['Foreign Gift Amount'].sum()
giftor_totals = giftor_totals.sort_values(ascending=False)
print('Top giftors by total amount:')
print(giftor_totals.head(10))

# Count how many gifts came from each giftor.
giftor_counts = known_giftors['Giftor Name'].value_counts()
print('\nTop giftors by number of gifts:')
print(giftor_counts.head(10))

Top giftors by total amount:
Giftor Name
Qatar Foundation                       1166503744
Qatar Foundation/Qatar National Res     796197000
Qatar Foundation for Education          373945215
Anonymous                               338793629
Saudi Arabian Cultural Mission          275221475
HCL                                     190000000
Church of Jesus Christ of LDS           185203715
Emirates Institute for Advanced Sc      170641244
QIC                                     148355497
Anonymous #9                             96334996
Name: Foreign Gift Amount, dtype: int64

Top giftors by number of gifts:
Giftor Name
Saudi Arabian Cultural Mission    480
Human Frontier Science Program    287
Anonymous                         262
Qatar Foundation                  261
Novartis Pharmaceuticals Corp     251
Royal Embassy of Saudi Arabia     210
Acerta Pharma B.V.                210
Contracting Party                 143
Gwangju Institute of S&T          141
WSB Universities                

**Answer:** By total amount, the leading named giftor is **Qatar Foundation** (**$1,166,503,744**). By count, it is **Saudi Arabian Cultural Mission** (**480 gifts**). All recipients in this education dataset are U.S. institutions, so no recipient-country filter is needed.